### ISSUE TITLE
Potential fails in installation or upgrade with corrupted XML file(s)

### ISSUE THREAD

**Issue by yuliu** [mybb/mybb#3770]
A potential incomplete (can be regarded as failed) MyBB installation/upgrade can appear, if the installation/upgrade is executed with corrupted XML files, especially `settings.xml` for board settings. OK, first the details, then comments.

MyBB uses XML files for structured data such as `settings`, `tasks`, `adminviews`, in both install and upgrade process, for flexibility (I suppose). The install/upgrade script doesn't check if an XML file is corrupted, either incomplete or malformatted. If the content of a XML file is wrong, the install/upgrade script might only output a warning (through foreach), thus we might have an failed installation with empty tables, eg. `settings` and `settinggroups`.

PHP file's corruption can be detected and reported by PHP, but the detect of corruption on XML files seems hard. I suggest at least put a md5 hash of each XML file in install/upgrade scripts to ensure its integrity. If a hash is not match, the installer/upgrader should inform user.

Dates back to [2015](https://community.mybb.com/thread-174046.html), there are users experiencing failed installation. Till today, we still have user reporting his incomplete installation of no settings in ACP > Configuration > Settings (via Discord Server). Although it appears that some of the uploaded MyBB files may be corrupted (not sure if any XML file involves).

MyBB may not be blamed, because the users are responsible for integrity of uploaded files. However, there can be various conditions that cause a file to be corrupted, network link error, transmission error or even server storage has bad blocks. And it could be more informative and friendly to users that the XML files are checked and a warning is issued if any XML file is incorrect during installation.

Sorry for a long story, I just try to explain it well.

**Comment by dvz** [comment]
Yes, this should be resolved when all files are verified on install and update: https://community.mybb.com/thread-214718.html (not on GitHub yet, I believe)

**Comment by yuliu** [comment]
If chances of getting corrupted are equal to a PHP file and an XML file, a corrupted PHP script or parsing a corrupted text file eg. `.sql` has a higher possibility of generating an error than an XML has. It's the nature of XML, which tends to be tolerable, that MyBB relies on. I've viewed SMF, phpBB, Discuz! and they use plain text files storing database structure and data rows to be populated. They parse these text files in their own way.

> Yes, this should be resolved when all files are verified on install and update: https://community.mybb.com/thread-214718.html (not on GitHub yet, I believe)

Sorry I missed it, I didn't search in the community forum for that topic. A pretty old problem 🤕. So, there could be two levels of integrity check before install/upgrade:
1. Install/Upgrade scripts and related files, including XML files, mainly under `./install/`.
1. The whole MyBB files, including PHP scripts, XML files, static html/css/js/images.

I suggest at least check those XML files during installation and upgrade process, which ensure or has its best effort to ensure the relevant database tables are correctly set up. As for other files, MyBB may release a utility script under `./install` folder, just as @dvz suggested, for [offline file integrity check](https://community.mybb.com/thread-214718-post-1334352.html#pid1334352).

**Comment by Ben-MyBB** [comment]
@dvz did you fix this one in the upgrade script?

**Comment by dvz** [comment]
1.9 install & upgrade run file verification with local checksums when in production environment: https://github.com/mybb/mybb/blob/4f0243ff2e4048d879b42cd0bee71f8e89d69922/inc/src/Maintenance/Process/Model/InstallModel.php#L49-L59
